server/src/main/kotlin/tech/figure/objectstore/gateway/components/BlockStreamConsumer.kt

```diff
@@ -0,0 +1,106 @@
+package tech.figure.objectstore.gateway.components
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import mu.KLogging
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.boot.actuate.health.Health
+import org.springframework.boot.actuate.health.HealthIndicator
+import org.springframework.boot.context.event.ApplicationReadyEvent
+import org.springframework.context.ApplicationListener
+import org.springframework.stereotype.Component
+import tech.figure.block.api.client.BlockAPIClient
+import tech.figure.block.api.proto.BlockServiceOuterClass
+import tech.figure.objectstore.gateway.configuration.BeanQualifiers
+import tech.figure.objectstore.gateway.configuration.BlockapiProperties
+import tech.figure.objectstore.gateway.repository.BlockHeightRepository
+import tech.figure.objectstore.gateway.service.TxEventHandlerService
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.ExperimentalTime
+
+@Component
+class BlockStreamConsumer(
+    private val txEventHandlerService: TxEventHandlerService,
+    private val blockHeightRepository: BlockHeightRepository,
+    private val blockapiProperties: BlockapiProperties,
+    private val blockAPIClient: BlockAPIClient,
+    @Qualifier(BeanQualifiers.BLOCK_STREAM_COROUTINE_SCOPE_QUALIFIER) private val blockStreamScope: CoroutineScope,
+) : ApplicationListener<ApplicationReadyEvent>, HealthIndicator {
+    private companion object : KLogging() {
+        private var blockStreamRunning = AtomicBoolean(false)
+        private val TX_EVENTS = setOf("wasm")
+    }
+
+    override fun onApplicationEvent(event: ApplicationReadyEvent) {
+        if (!blockapiProperties.enabled) {
+            logger.warn("Event stream has been manually disabled! Use a value of `true` for EVENT_STREAM_ENABLED to re-enable it")
+            return
+        }
+        tryStartBlockStream {
+            blockStreamLoop()
+        }
+    }
+
+    override fun health(): Health = if (blockStreamRunning.get()) {
+        Health.up().build()
+    } else {
+        Health.down().build()
+    }
+
+    private suspend fun blockStreamLoop() {
+        val lastBlockProcessed = blockHeightRepository.getLastProcessedBlockHeight()
+        blockAPIClient.streamBlocks(
+            start = lastBlockProcessed,
+            preference = BlockServiceOuterClass.PREFER.TX_EVENTS
+        ).collect { block ->
+            val lastProcessedHeight = blockHeightRepository.getLastProcessedBlockHeight()
+
+            block.blockResult.block.transactionsList.flatMap { it.eventsList }
```

Comment:
took me a while to figure out that block.blockResult.transactions.transactions.transactionsList was an empty red herring 🤦

Reply:
Yea, there's a branch out there to nix that out that's about 8 months old.